### ISSUE TITLE
Completely separated client and server sides?

### ISSUE THREAD

**Issue by frgomes** [ericmdantas/generator-ng-fullstack#247]
I suggest splitting apart server and client sides, making them completely separate and independent directory structures.

Optionally, when the user specifies *both*, a top level build script could provide shortcuts for automating mundane tasks performed on both sides, like ``npm install`` on both sides; ``ng serve`` on the client side and ``gulp run-watch`` on the server side, etc.

The user could then choose to run build/run scripts separately when convenient or run everything in one go when convenient, depending on specific needs or preferences.

**Comment by ericmdantas** [comment]
What kind of separation are you suggesting?

For now we already have the `client` and `server` folders separated. The only thing both structures have to know from each other is the path for the dependencies - so the server can serve the static content.

**Comment by frgomes** [comment]
Directories ``client`` and ``server`` are not complete projects, but only typescript sources for the application, among other things. These directories do not contain build scripts, for example.

I would like the ability to do change the build process or other characteristics of the server side, for example, without having to touch anything which could potentially break the client side.

The need I initially had was the ability to reload the code automagically when sources get changed. This works well on the client side but not on the server side. Researching over the internet, I found projects able to do that. In particular, I found one project which is only the server side, no frontend at all.

Back to ``ng-fullstack``, we are employing ``ng serve`` for monitoring and reloading the application when it gets changed, which is OK. However, I would like to transplant the ``gulpfile`` I found on the other project to the server side of ``ng-fullstack``. But there's no really "server side of ng-fullstack"... it's only one single project, with only one build script (even though there's a ``tasks`` directory, etc).

In a nutshell, I would like to have separate build scripts for each side, which would allow me to say ``ng serve`` on the ``client`` side and say ``gulp watch-run`` on the ``server`` side.

For your information, I'm building server and client side completely separated from each other, assembly bits and pieces produces by ``ng-fullstack`` and the other project I've mentioned.

Yes, I understand that client and server share the knowledge of the data being transferred between them... and possibly this would require a shared folder where interfaces are defined, for example. But I'm leaving this problem for a future iteration.

Thanks

**Comment by ericmdantas** [comment]
Oh, I see what you mean. Some things might be coupled, but it shouldn't be too compliacated to adapt/customize tasks, among other stuff. The separations of folders were created with that in mind.

Obviously, integrating with some projects will be easier than others, but that's expected.

Just a few notes:

> Directories client and server are not complete projects, but only typescript sources for the application, among other things. These directories do not contain build scripts, for example.

The `tasks` folder have those.

> I would like the ability to do change the build process or other characteristics of the server side

Inside `tasks` there's the `server` folder, which contains only the server stuff, not related to the client side.

> The need I initially had was the ability to reload the code automagically

You can use `nodemon` for that, it's already working on `v1.9`, but it's not released yet; see the line in the `package.json` here: https://github.com/ericmdantas/generator-ng-fullstack/blob/v1.9/app/templates/_package.json#L16

> However, I would like to transplant the gulpfile I found on the other project to the server side of ng-fullstack

I guess that could be done with the change only being done in `tasks/server`.

**Comment by frgomes** [comment]
I'm using only _npm_ now as per this article:
https://www.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/

I have now a completely self contained frontend and a completely self contained backend. My main project has only a .gitmodules file and that's all.

In particular, I've removed _gulp_ from the build process. The reason for that is that now I'm using a tool called _s2i_ from _RedHat OpenShift_, which builds the application and generates a Docker image which conforms with best practices. I've found difficulties running _s2i_ because, in a nutshell, _gulp_ was not installed in the default bootstrap environment, which contains only _node_ and _npm_. For this reason, I've completely eliminated the need of _gulp_ in my build process.

For our information, this is how the build process of the backend looks like:

``` json
  "devDependencies": {
    "watch": "1.0.1",
    "nodemon": "1.11.0",
    "parallelshell": "2.0.0",
    "typescript": "^2.1.5",
    "ntypescript": "1.201609302242.1",
    "typings": "^2.1.0",
    "tslint": "^4.4.2",
    "mocha": "^2.4.5",
    "should": "^11.2.0"
  },

  "scripts": {
    "postinstall": "npm run build",
    
    "start":       "node dist/server.js",
    
    "clean":   "rimraf dist",
    "zap":     "rimraf dist node_modules typings",

    "build":   "typings install && npm run compile && npm run test",
    "compile": "ntsc --outDir ./dist",

    "test":    "mocha",

    "watch": "watch 'npm run build' src",
    "reload": "nodemon --watch dist",
    "watch:reload": "parallelshell 'npm run watch' 'npm run reload'",

    "tools:typings": "npm install -g typings@2.1.0 && typings install"
  },
```

Doing this way, the backend can be reloaded when sources change, employing _npm_ only.

I hope I've contributed with something useful.
Thanks a lot for your great work on ng-fullstack! :-)

**Comment by ericmdantas** [comment]
Thank you for all the information, @frgomes, it's definitely useful - and it's also a good example for someone who wants to do something like in your case.

👍
